File: SDE_sheet/Linked-List/04_remove_nth_node_from_list/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* dummy = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummy->val = 0;
    dummy->next = head;

    struct ListNode* slow = dummy;
    struct ListNode* fast = dummy;

    for (int i = 0; i <= n; i++) {
        fast = fast->next;
    }

    while (fast != NULL) {
        slow = slow->next;
        fast = fast->next;
    }

    struct ListNode* toDelete = slow->next;
    slow->next = slow->next->next;
    free(toDelete);

    struct ListNode* newHead = dummy->next;
    free(dummy);

    return newHead;
}

File: SDE_sheet/Linked-List/04_remove_nth_node_from_list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    const dummy: ListNode = new ListNode(0, head);
    let slow: ListNode = dummy;
    let fast: ListNode = dummy;

    for (let i = 0; i < n; i++) {
        fast = fast.next!;
    }

    while (fast.next) {
        slow = slow.next!;
        fast = fast.next;
    }

    slow.next = slow.next!.next;
    return dummy.next;
};
